### ISSUE TITLE
Circular buffer data structure

### ISSUE THREAD

**Issue by DixonD-git** [dotnet/corefx#3126]
As far as I know, there is no circular buffer implementation in .NET (https://en.wikipedia.org/wiki/Circular_buffer). Some simple implementations over an array or a Queue are discussed here, for instance: https://social.msdn.microsoft.com/forums/vstudio/en-US/416a2175-b05d-43b1-b99a-a01c56550dbe/circular-buffer-in-net

**Comment by davidfowl** [comment]
We have something like this in SignalR https://github.com/aspnet/SignalR-Server/blob/dev/src/Microsoft.AspNet.SignalR.Messaging/MessageStore.cs

**Comment by svick** [comment]
So, the difference between a circular buffer and `Queue<T>` is that when you try to go over the capacity, circular buffer overwrites items or indicates a failure (which one?), whereas `Queue<T>` resizes itself?

To me, that sounds like this might be simple enough and/or not general enough to be included.

**Comment by GrabYourPitchforks** [comment]
@svick generally it overwrites the oldest items. The typical use case is for something like a log which contains only the last xyz number of entries, and as new entries are added they'll evict the oldest entries.

The data structure isn't really difficult to implement, but there are some nits around efficiency and keeping indexes in sync. From a "we'll do the grunt work of implementing this so our consumers can focus on being productive" library author's perspective, it may be interesting.

The particular data structure used by SignalR is highly optimized for their unique scenario and access patterns, so I wouldn't use it as the design basis for a general-purpose framework type.

**Comment by davidfowl** [comment]
This is also interesting http://lmax-exchange.github.io/disruptor/

**Comment by mattwarren** [comment]
@davidfowl there's a .NET port of disruptor available, see https://github.com/disruptor-net/Disruptor-net

**Comment by karelz** [comment]
We need formal API proposal.

**Comment by SunnyWar** [comment]
I skeptical how needed this is. Writing one is trivial by using a Queue and called dequeuer if the size is reached before pushing another item. I suppose someone might want one with vastly superior performance than what is described above, but unless such a thing is demonstrated, I'm not seeing the value.  In short, show me numbers that demonstrate a circular buffer that vastly outperforms the trivial implementation and I'll be 100% behind it.

**Comment by karelz** [comment]
I see the value mostly in the "I don't have to write the trivial implementation".

**Comment by benaadams** [comment]
The [`Segment`](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentQueue.cs#L849-L1139) in `ConcurrentQueue<T>` is probably already quite close and high performance; but would need to overwrite rather than returning false.

No one uses non-power of two sized ring buffers do they? 😉

**Comment by Drawaes** [comment]
Disruptor ... be wary it's not as fast as you might expect because it covers many different scenarios . It turns out it's a generalist ..

**Comment by Drawaes** [comment]
Are you saying you want a threadsafe circular buffer ?

**Comment by pgolebiowski** [comment]
> Are you saying you want a threadsafe circular buffer?

@Drawaes Actually, it has happened to me that I needed a concurrent circular buffer once. 😄 What do you guys think about adding it?

- Casual circular buffer would satisfy @karelz -> _"I don't have to write the trivial implementation."_
- Concurrent one would satisfy @SunnyWar -> _"a circular buffer that vastly outperforms the trivial implementation and I'll be 100% behind it."_
- Implementing both would satisfy me. 😄

**Comment by karelz** [comment]
Not sure if adding both isn't overkill. The concurrent would be best to come with data as discussed above.

@pgolebiowski if you want to help move .NET Core 2.0 forward where it needs it most, please check our announcement: #17619.

**Comment by Drawaes** [comment]
Non threadsafe circular buffer is a couple of lines so in my opinion not worth an corefx push. But if there is a need for a concurrent one then sure go ahead and add it. Anytime I have needed one the program control is single threaded by that point so the naive simple case (array mod the index) has been all I have looked at.

**Comment by benaadams** [comment]
Generic version

```csharp
namespace System.Collections.Generic
{
    /// <summary>
    /// Represents a bounded first-in, first-out collection of objects.
    /// </summary>
    /// <typeparam name="T">Specifies the type of elements in the queue.</typeparam>    
    [DebuggerDisplay("Count = {count}")]
    [DebuggerTypeProxy(typeof(System_CircularQueueDebugView<>))]
    [Serializable]
    public class CircularQueue<T> : IEnumerable<T>, IReadOnlyCollection<T>
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="CircularQueue{T}"/> class 
        /// that has the specified maximum capacity.
        /// </summary>
        /// <param name="capacity">The number of elements that the <see cref="CircularQueue{T}"/> can contain.</param>
        /// <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="capacity"/> is less than zero.</exception>
        public CircularQueue(int capacity);

        /// <summary>
        /// Initializes a new instance of the <see cref="CircularQueue{T}"/> class 
        /// that contains elements copied from the specified collection and maximum capacity set to <paramref name="collection"/> Count.
        /// </summary>
        /// <param name="collection">The collection whose elements are copied to the new <see cref="CircularQueue{T}"/>.</param>
        /// <exception cref="T:System.ArgumentNullException"><paramref name="collection"/> is null.</exception>
        public CircularQueue(IEnumerable<T> collection);

        /// <summary>
        /// Initializes a new instance of the <see cref="CircularQueue{T}"/> class 
        /// that contains elements copied from the specified collection and has the specified maximum capacity.
        /// </summary>
        /// <param name="collection">The collection whose elements are copied to the new <see cref="CircularQueue{T}"/>.</param>
        /// <exception cref="T:System.ArgumentNullException"><paramref name="collection"/> is null.</exception>
        /// <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="capacity"/> 
        /// is less than zero or less than <paramref name="collection"/> count.</exception>
        public CircularQueue(IEnumerable<T> collection, int capacity);
        
        /// <summary>
        /// Gets the number of elements contained in the <see cref="CircularQueue{T}"/>.
        /// </summary>
        /// <value>The number of elements contained in the <see cref="CircularQueue{T}"/>.</value>
        public int Count { get; }

        /// <summary>
        /// Gets a value that indicates whether the <see cref="CircularQueue{T}"/> is empty.
        /// </summary>
        /// <value>true if the <see cref="CircularQueue{T}"/> is empty; otherwise, false.</value>
        /// <remarks>
        public bool IsEmpty { get; }
        
        /// <summary>
        /// Attempts to add an object to the end of the <see cref="CircularQueue{T}"/>, 
        /// if the queue is full it returns false.
        /// </summary>
        /// <param name="item">
        /// The object to add to the <see cref="CircularQueue{T}"/>. 
        /// The value can be null for reference types.
        /// </param>
        public bool TryEnqueue(T item);

        /// <summary>
        /// Adds an object to the end of the <see cref="CircularQueue{T}"/>, 
        /// if the full the oldest item is overwriten.
        /// </summary>
        /// <param name="item">
        /// The object to add to the <see cref="CircularQueue{T}"/>. 
        /// The value can be null for reference types.
        /// </param>
        public void Enqueue(T item);

        /// <summary>
        /// Attempts to remove and return the object at the beginning of the <see cref="CircularQueue{T}"/>.
        /// </summary>
        /// <param name="result">
        /// When this method returns, if the operation was successful, <paramref name="result"/> contains the
        /// object removed. If no object was available to be removed, the value is unspecified.
        /// </param>
        /// <returns>
        /// true if an element was removed and returned from the beginning of the
        /// <see cref="CircularQueue{T}"/> successfully; otherwise, false.
        /// </returns>
        public bool TryDequeue(out T result);

        /// <summary>
        /// Attempts to return the object at the beginning of the <see cref="CircularQueue{T}"/> 
        /// without removing it.
        /// </summary>
        /// <param name="result">
        /// When this method returns, <paramref name="result"/> contains an object from
        /// the beginning of the <see cref="CircularQueue{T}"/> or default(T)
        /// if the operation failed.
        /// </param>
        /// <returns>true if and object was returned successfully; otherwise, false.</returns>
        /// <remarks>
        /// For determining whether the collection contains any items, use of the <see cref="IsEmpty"/>
        /// property is recommended rather than peeking.
        /// </remarks>
        public bool TryPeek(out T result);

        /// <summary>
        /// Removes all elements from the <see cref="CircularQueue{T}"/>.
        /// </summary>
        public void Clear();

        /// <summary>
        /// Returns an enumerator that iterates through the <see cref="CircularQueue{T}"/>
        /// </summary>
        /// <returns>An enumerator for the contents of the <see cref="CircularQueue{T}"/>.</returns>
        public Enumerator GetEnumerator();

        /// <summary>
        /// Returns an enumerator that iterates through the <see cref="CircularQueue{T}"/>
        /// </summary>
        /// <returns>An enumerator for the contents of the <see cref="CircularQueue{T}"/>.</returns>
        IEnumerator<T> IEnumerable<T>.GetEnumerator();

        /// <summary>
        /// Returns an enumerator that iterates through the <see cref="CircularQueue{T}"/>.
        /// </summary>
        /// <returns>An <see cref="T:System.Collections.IEnumerator"/> that can be used to iterate 
        /// through the collection.</returns>
        IEnumerator IEnumerable.GetEnumerator();

        public struct Enumerator : IEnumerator<T>
        {
            public T Current { get; }
            public bool MoveNext();
            public void Dispose();
            object IEnumerator.Current { get; }
            void IEnumerator.Reset();
        }
    }
}
```

**Comment by benaadams** [comment]
Concurrent version 

```csharp
namespace System.Collections.Concurrent
{
    /// <summary>
    /// Represents a thread-safe bounded first-in, first-out collection of objects.
    /// </summary>
    /// <typeparam name="T">Specifies the type of elements in the queue.</typeparam>    
    /// <remarks>
    /// All public and protected members of <see cref="ConcurrentCircularQueue{T}"/> are thread-safe and may be used
    /// concurrently from multiple threads.
    /// </remarks>
    [DebuggerDisplay("Count = {count}")]
    [DebuggerTypeProxy(typeof(System_ConcurrentCircularQueueDebugView<>))]
    [Serializable]
    public class ConcurrentCircularQueue<T> : IEnumerable<T>, IReadOnlyCollection<T>
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="ConcurrentCircularQueue{T}"/> class 
        /// that has the specified maximum capacity.
        /// </summary>
        /// <param name="capacity">The number of elements that the <see cref="ConcurrentCircularQueue{T}"/> can contain.</param>
        /// <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="capacity"/> is less than zero.</exception>
        public ConcurrentCircularQueue(int capacity);

        /// <summary>
        /// Initializes a new instance of the <see cref="ConcurrentCircularQueue{T}"/> class 
        /// that contains elements copied from the specified collection and maximum capacity set to <paramref name="collection"/> Count.
        /// </summary>
        /// <param name="collection">The collection whose elements are copied to the new <see cref="ConcurrentCircularQueue{T}"/>.</param>
        /// <exception cref="T:System.ArgumentNullException"><paramref name="collection"/> is null.</exception>
        public ConcurrentCircularQueue(IEnumerable<T> collection);

        /// <summary>
        /// Initializes a new instance of the <see cref="ConcurrentCircularQueue{T}"/> class 
        /// that contains elements copied from the specified collection and has the specified maximum capacity.
        /// </summary>
        /// <param name="collection">The collection whose elements are copied to the new <see cref="ConcurrentCircularQueue{T}"/>.</param>
        /// <exception cref="T:System.ArgumentNullException"><paramref name="collection"/> is null.</exception>
        /// <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="capacity"/> 
        /// is less than zero or less than <paramref name="collection"/> count.</exception>
        public ConcurrentCircularQueue(IEnumerable<T> collection, int capacity);
        
        /// <summary>
        /// Gets the number of elements contained in the <see cref="ConcurrentCircularQueue{T}"/>.
        /// </summary>
        /// <value>The number of elements contained in the <see cref="ConcurrentCircularQueue{T}"/>.</value>
        /// <remarks>
        /// For determining whether the collection contains any items, use of the <see cref="IsEmpty"/>
        /// property is recommended rather than retrieving the number of items from the <see cref="Count"/>
        /// property and comparing it to 0.
        /// </remarks>
        public int Count { get; }

        /// <summary>
        /// Gets a value that indicates whether the <see cref="ConcurrentCircularQueue{T}"/> is empty.
        /// </summary>
        /// <value>true if the <see cref="ConcurrentCircularQueue{T}"/> is empty; otherwise, false.</value>
        /// <remarks>
        /// For determining whether the collection contains any items, use of this property is recommended
        /// rather than retrieving the number of items from the <see cref="Count"/> property and comparing it
        /// to 0.  However, as this collection is intended to be accessed concurrently, it may be the case
        /// that another thread will modify the collection after <see cref="IsEmpty"/> returns, thus invalidating
        /// the result.
        /// </remarks>
        public bool IsEmpty { get; }
        
        /// <summary>
        /// Attempts to add an object to the end of the <see cref="ConcurrentCircularQueue{T}"/>, 
        /// if the queue is full it returns false.
        /// </summary>
        /// <param name="item">
        /// The object to add to the <see cref="ConcurrentCircularQueue{T}"/>. 
        /// The value can be null for reference types.
        /// </param>
        public bool TryEnqueue(T item);

        /// <summary>
        /// Adds an object to the end of the <see cref="ConcurrentCircularQueue{T}"/>, 
        /// if the full the oldest item is overwriten.
        /// </summary>
        /// <param name="item">
        /// The object to add to the <see cref="ConcurrentCircularQueue{T}"/>. 
        /// The value can be null for reference types.
        /// </param>
        public void Enqueue(T item);

        /// <summary>
        /// Attempts to remove and return the object at the beginning of the <see cref="ConcurrentCircularQueue{T}"/>.
        /// </summary>
        /// <param name="result">
        /// When this method returns, if the operation was successful, <paramref name="result"/> contains the
        /// object removed. If no object was available to be removed, the value is unspecified.
        /// </param>
        /// <returns>
        /// true if an element was removed and returned from the beginning of the
        /// <see cref="ConcurrentCircularQueue{T}"/> successfully; otherwise, false.
        /// </returns>
        public bool TryDequeue(out T result);

        /// <summary>
        /// Attempts to return the object at the beginning of the <see cref="ConcurrentCircularQueue{T}"/> 
        /// without removing it.
        /// </summary>
        /// <param name="result">
        /// When this method returns, <paramref name="result"/> contains an object from
        /// the beginning of the <see cref="ConcurrentCircularQueue{T}"/> or default(T)
        /// if the operation failed.
        /// </param>
        /// <returns>true if and object was returned successfully; otherwise, false.</returns>
        /// <remarks>
        /// For determining whether the collection contains any items, use of the <see cref="IsEmpty"/>
        /// property is recommended rather than peeking.
        /// </remarks>
        public bool TryPeek(out T result);

        /// <summary>
        /// Removes all elements from the <see cref="ConcurrentCircularQueue{T}"/>.
        /// </summary>
        public void Clear();

        /// <summary>
        /// Returns an enumerator that iterates through the <see cref="ConcurrentCircularQueue{T}"/>
        /// </summary>
        /// <returns>An enumerator for the contents of the <see cref="ConcurrentCircularQueue{T}"/>.</returns>
        public Enumerator GetEnumerator();

        /// <summary>
        /// Returns an enumerator that iterates through the <see cref="ConcurrentCircularQueue{T}"/>
        /// </summary>
        /// <returns>An enumerator for the contents of the <see cref="ConcurrentCircularQueue{T}"/>.</returns>
        IEnumerator<T> IEnumerable<T>.GetEnumerator();

        /// <summary>
        /// Returns an enumerator that iterates through the <see cref="ConcurrentCircularQueue{T}"/>.
        /// </summary>
        /// <returns>An <see cref="T:System.Collections.IEnumerator"/> that can be used to iterate 
        /// through the collection.</returns>
        IEnumerator IEnumerable.GetEnumerator();

        public struct Enumerator : IEnumerator<T>
        {
            public T Current { get; }
            public bool MoveNext();
            public void Dispose();
            object IEnumerator.Current { get; }
            void IEnumerator.Reset();
        }
    }
}
```

**Comment by benaadams** [comment]
`Enqueue` overwrites and `TryEnqueue` doesn't

**Comment by khellang** [comment]
I think the concurrent one should implement `IProducerConsumerCollection<T>` :smile:

**Comment by C-Babbage** [comment]
I’m new to contributing, but I’d be happy to give this a go if no one else has claimed it

**Comment by khellang** [comment]
@C-Babbage The [process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) needs a formal API proposal (something like what @benaadams provided) and the issue needs to be marked as `api-ready-for-review` and go through the review process before there's any point in sending a pull request.

You can, off course, start working on it before submitting an API proposal.

**Comment by Drawaes** [comment]
I would take a look at the new specialized collections package being worked on in corefxlab. You will avoid the heavy API being added to .net core process and get eyes on. If it is useful and heavily used it could then be proposed for corefx

**Comment by C-Babbage** [comment]
@Drawaes that seems like a good idea. That package is fairly recent too. Should we close this one out and move the proposal?

**Comment by oising** [comment]
> I think the concurrent one should implement [`IProducerConsumerCollection<T>`](https://msdn.microsoft.com/en-us/library/dd287147(v=vs.110).aspx), perhaps explicitly 😄

I came here looking to find one, so I didn't have to write my own "trivial implementation." In particular, I want a thread safe blocking circular buffer/queue. This would be satisfied if the above was implemented, so I could pass it to BlockingCollection<T>.

**Comment by oising** [comment]
https://raw.githubusercontent.com/helios-io/helios/5313bc4b52d0cb81227551b520b440aff21c1764/src/Helios/Util/Collections/ConcurrentCircularBuffer.cs

**Comment by SunnyWar** [comment]
@oising In about that time it took you to search for and then write a post about it, you could have written it. This is why I'm spectacle about the need for it in the framework. Also, there are too many design decisions that are unclear and making them will force the implementation in a direction that might no suite the needs. For example, typically a circular queue has a fixed capacity. If the constructor takes a collection, should the capacity be set to the number of items in the collection or use whatever default value it has? Or perhaps there should be no default constructor and hence no default capacity and instead it must be specified in the constructor. Also, when and item is Enqueue'd and an item that is in the queue needs to be discarded, how do you notify that caller that this happened? Or should it even bother? Some use cases won't care and others will care a great deal. 
In short a "simple" circularqueue is trivial to write (inherit or encapsulate a queue or concurrentqueue, add a Capacity property and override or "new" the Enqueue method) BUT going beyond that seems to be so specific to the particular need of the application I know of no way to truly cover 80% of the needs.

**Comment by oising** [comment]
@SunnyWar You're quite right. It's surprising how deep this rabbit hole can go:

https://github.com/Azure/DotNetty/blob/dev/src/DotNetty.Common/Internal/ConcurrentCircularArrayQueue.cs

**Comment by oising** [comment]
Huh... how come I didn't see this before? Won't this do the trick? 

https://docs.microsoft.com/en-us/dotnet/api/system.threading.channels.channel.createbounded

A bounded channel with "full" semantics of "drop oldest" seems to approximate a circular buffer. It can be threadsafe for readers and/or writers with some flags. Well, I guess the framework already has an implementation :)

**Comment by SunnyWar** [comment]
@oising I've been curious about the Channel object for awhile so I took this opportunity to try it. I wrote a very simple performance tests comparing it with a simple CircularQueue using a Queue as a backing store. "add" simply adds a few million items. "read" simply adds one and the reads one item.
00:00:01.1691329 CircularQueue add
00:00:03.7712431 Channel add
00:00:01.2131137 CircularQueue read
00:00:08.3058750 Channel read
It's not worth it to me to continue looking into this Channel object as it's so much slower.

**Comment by oising** [comment]
Did you test ensure that SingleReader and Single Writer options are set to
true, to approximate your non threadsafe (I assume) implementation? Channel
défauts to threadsafe read AND write, which is bound (har har) to slow
everything down

**Comment by GSPP** [comment]
Maybe it is better to add a `Deque` and use that as a ring buffer.

```
static void RingBufferStyleAdd(this Deque<T> d, T item) {
 if (d.Count == d.Capacity) d.PopEnd();
 d.PushFront(item);
}
```

I am in favor of adding a deque to the framework (https://github.com/dotnet/corefx/issues/32790).

**Comment by stephentoub** [comment]
> Maybe it is better to add a Deque and use that as a ring buffer.

Why is using a `Deque<T>` for this any better than using a `Queue<T>` for this?

**Comment by GSPP** [comment]
@stephentoub That's a good point. I got confused there.

That makes it even less attractive to have a specialized ring buffer structure in the framework.
